File: matlab/Tutorials/Horn_Antenna.m
%
% Tutorials / horn antenna
%
% Describtion at:
% http://openems.de/index.php/Tutorial:_Horn_Antenna
%
% Tested with
%  - Matlab 2011a / Octave 3.6.4
%  - openEMS v0.0.31
%
% (C) 2011,2012,2013 Thorsten Liebig <thorsten.liebig@uni-due.de>

close all
clear
clc

%% setup the simulation
physical_constants;
unit = 1e-3; % all length in mm

% horn width in x-direction
horn.width  = 20;
% horn height in y-direction
horn.height = 30;
% horn length in z-direction
horn.length = 50;

horn.feed_length = 50;

horn.thickness = 2;

% horn opening angle in x, y
horn.angle = [20 20]*pi/180;

% size of the simulation box
SimBox = [200 200 200];

% frequency range of interest
f_start =  10e9;
f_stop  =  20e9;

% frequency of interest
f0 = 15e9;

%waveguide TE-mode definition
TE_mode = 'TE10';
a = horn.width;
b = horn.height;

%% setup FDTD parameter & excitation function
FDTD = InitFDTD('EndCriteria', 1e-4);
FDTD = SetGaussExcite(FDTD,0.5*(f_start+f_stop),0.5*(f_stop-f_start));
BC = {'PML_8' 'PML_8' 'PML_8' 'PML_8' 'PML_8' 'PML_8'}; % boundary conditions
FDTD = SetBoundaryCond( FDTD, BC );

%% setup CSXCAD geometry & mesh
% currently, openEMS cannot automatically generate a mesh
max_res = c0 / (f_stop) / unit / 15; % cell size: lambda/20
CSX = InitCSX();

%create fixed lines for the simulation box, substrate and port
mesh.x = [-SimBox(1)/2 -a/2 a/2 SimBox(1)/2];
mesh.x = SmoothMeshLines( mesh.x, max_res, 1.4); % create a smooth mesh between specified fixed mesh lines

mesh.y = [-SimBox(2)/2 -b/2 b/2 SimBox(2)/2];
mesh.y = SmoothMeshLines( mesh.y, max_res, 1.4 );

%create fixed lines for the simulation box and given number of lines inside the substrate
mesh.z = [-horn.feed_length 0 SimBox(3)-horn.feed_length ];
mesh.z = SmoothMeshLines( mesh.z, max_res, 1.4 );

CSX = DefineRectGrid( CSX, unit, mesh );

%% create horn
% horn feed rect waveguide
CSX = AddMetal(CSX, 'horn');
start = [-a/2-horn.thickness -b/2 mesh.z(1)];
stop  = [-a/2                 b/2 0];
CSX = AddBox(CSX,'horn',10,start,stop);
start = [a/2+horn.thickness -b/2 mesh.z(1)];
stop  = [a/2                 b/2 0];
CSX = AddBox(CSX,'horn',10,start,stop);
start = [-a/2-horn.thickness b/2+horn.thickness mesh.z(1)];
stop  = [ a/2+horn.thickness b/2                0];
CSX = AddBox(CSX,'horn',10,start,stop);
start = [-a/2-horn.thickness -b/2-horn.thickness mesh.z(1)];
stop  = [ a/2+horn.thickness -b/2                0];
CSX = AddBox(CSX,'horn',10,start,stop);

% horn opening
p(2,1) = a/2;
p(1,1) = 0;
p(2,2) = a/2 + sin(horn.angle(1))*horn.length;
p(1,2) = horn.length;
p(2,3) = -a/2 - sin(horn.angle(1))*horn.length;
p(1,3) = horn.length;
p(2,4) = -a/2;
p(1,4) = 0;
CSX = AddLinPoly( CSX, 'horn', 10, 1, -horn.thickness/2, p, horn.thickness, 'Transform', {'Rotate_X',horn.angle(2),'Translate',['0,' num2str(-b/2-horn.thickness/2) ',0']});
CSX = AddLinPoly( CSX, 'horn', 10, 1, -horn.thickness/2, p, horn.thickness, 'Transform', {'Rotate_X',-horn.angle(2),'Translate',['0,' num2str(b/2+horn.thickness/2) ',0']});

p(1,1) = b/2+horn.thickness;
p(2,1) = 0;
p(1,2) = b/2+horn.thickness + sin(horn.angle(2))*horn.length;
p(2,2) = horn.length;
p(1,3) = -b/2-horn.thickness - sin(horn.angle(2))*horn.length;
p(2,3) = horn.length;
p(1,4) = -b/2-horn.thickness;
p(2,4) = 0;
CSX = AddLinPoly( CSX, 'horn', 10, 0, -horn.thickness/2, p, horn.thickness, 'Transform', {'Rotate_Y',-horn.angle(2),'Translate',[ num2str(-a/2-horn.thickness/2) ',0,0']});
CSX = AddLinPoly( CSX, 'horn', 10, 0, -horn.thickness/2, p, horn.thickness, 'Transform', {'Rotate_Y',+horn.angle(2),'Translate',[ num2str(a/2+horn.thickness/2) ',0,0']});

% horn aperture
A = (a + 2*sin(horn.angle(1))*horn.length)*unit * (b + 2*sin(horn.angle(2))*horn.length)*unit;

%% apply the excitation %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
start=[-a/2 -b/2 mesh.z(8) ];
stop =[ a/2  b/2 mesh.z(1)+horn.feed_length/2 ];
[CSX, port] = AddRectWaveGuidePort( CSX, 0, 1, start, stop, 2, a*unit, b*unit, TE_mode, 1);

%% nf2ff calc
start = [mesh.x(9) mesh.y(9) mesh.z(9)];
stop  = [mesh.x(end-8) mesh.y(end-8) mesh.z(end-8)];
[CSX nf2ff] = CreateNF2FFBox(CSX, 'nf2ff', start, stop, 'Directions', [1 1 1 1 0 1]);

%% prepare simulation folder
Sim_Path = 'tmp_Horn_Antenna';
Sim_CSX = 'horn_ant.xml';

[status, message, messageid] = rmdir( Sim_Path, 's' ); % clear previous directory
[status, message, messageid] = mkdir( Sim_Path ); % create empty simulation folder

%% write openEMS compatible xml-file
WriteOpenEMS([Sim_Path '/' Sim_CSX], FDTD, CSX);

%% show the structure
CSXGeomPlot([Sim_Path '/' Sim_CSX]);

%% run openEMS
RunOpenEMS(Sim_Path, Sim_CSX);

%% postprocessing & do the plots
freq = linspace(f_start,f_stop,201);

port = calcPort(port, Sim_Path, freq);

Zin = port.uf.tot ./ port.if.tot;
s11 = port.uf.ref ./ port.uf.inc;

plot( freq/1e9, 20*log10(abs(s11)), 'k-', 'Linewidth', 2 );
ylim([-60 0]);
grid on
title( 'reflection coefficient S_{11}' );
xlabel( 'frequency f / GHz' );
ylabel( 'reflection coefficient |S_{11}|' );

drawnow

%% NFFF contour plots %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

% calculate the far field at phi=0 degrees and at phi=90 degrees
thetaRange = (0:2:359) - 180;
disp( 'calculating far field at phi=[0 90] deg...' );
nf2ff = CalcNF2FF(nf2ff, Sim_Path, f0, thetaRange*pi/180, [0 90]*pi/180);

Dlog=10*log10(nf2ff.Dmax);
G_a = 4*pi*A/(c0/f0)^2;
e_a = nf2ff.Dmax/G_a;

% display some antenna parameter
disp( ['radiated power: Prad = ' num2str(nf2ff.Prad) ' Watt']);
disp( ['directivity: Dmax = ' num2str(Dlog) ' dBi'] );
disp( ['aperture efficiency: e_a = ' num2str(e_a*100) '%'] );

%%
% normalized directivity
figure
plotFFdB(nf2ff,'xaxis','theta','param',[1 2]);
drawnow
%   D_log = 20*log10(nf2ff.E_norm{1}/max(max(nf2ff.E_norm{1})));
%   D_log = D_log + 10*log10(nf2ff.Dmax);
%   plot( nf2ff.theta, D_log(:,1) ,'k-', nf2ff.theta, D_log(:,2) ,'r-' );

% polar plot
figure
polarFF(nf2ff,'xaxis','theta','param',[1 2],'logscale',[-40 20], 'xtics', 12);
drawnow
%   polar( nf2ff.theta, nf2ff.E_norm{1}(:,1) )

%% calculate 3D pattern
phiRange = sort( unique( [-180:5:-100 -100:2.5:-50 -50:1:50 50:2.5:100 100:5:180] ) );
thetaRange = sort( unique([ 0:1:50 50:2.:100 100:5:180 ]));

disp( 'calculating 3D far field...' );
nf2ff = CalcNF2FF(nf2ff, Sim_Path, f0, thetaRange*pi/180, phiRange*pi/180, 'Verbose',2,'Outfile','nf2ff_3D.h5');

figure
plotFF3D(nf2ff);

%%
E_far_normalized = nf2ff.E_norm{1}/max(nf2ff.E_norm{1}(:));
DumpFF2VTK([Sim_Path '/Horn_Pattern.vtk'],E_far_normalized,thetaRange,phiRange,'scale',1e-3);

File: python/Tutorials/Horn_Antenna.py
# -*- coding: utf-8 -*-
"""
 Horn Antenna Tutorial with Coaxial Pin Feed

 A pyramidal horn antenna fed by a coaxial probe (pin) inside the
 rectangular feed waveguide.  The back of the waveguide is closed by a
 metal short-circuit plate (back-short) instead of being terminated into a
 PML absorber.  A vertical pin, modelled as a lumped port, excites the TE10
 waveguide mode.

 Design rules used here:
   * Back-short distance from pin to the back wall = lambda_g / 4
     (quarter guided wavelength at f0 ⟹ maximum forward wave coupling)
   * Pin length  ≈ 0.55 * b_wg  (starting point; tune for 50 Ω match)
   * Pin is centred at x=0 where the TE10 E-field is maximum

 Tested with
  - python 3.14
  - openEMS v0.37

 (c) 2025 Thorsten Liebig <thorsten.liebig@gmx.de>

"""

### Import Libraries
import os, tempfile
import numpy as np
import matplotlib.pyplot as plt

from CSXCAD import ContinuousStructure
from openEMS import openEMS
from openEMS.physical_constants import *

### Simulation path
Sim_Path = os.path.join(tempfile.gettempdir(), 'Horn_Antenna_PinFeed')

post_proc_only = False

unit = 1e-3     # all lengths in mm

### Waveguide and Horn Parameters
a_wg = 20.0     # feed waveguide width  [mm] – TE10 propagates above fc = C0/(2a)
b_wg = 10.0     # feed waveguide height [mm]
wg_t  = 2.0     # metal wall thickness  [mm]

feed_length = 50.0  # length of the closed feed waveguide behind the horn mouth [mm]

horn_length = 50.0  # horn length in propagation direction (z) [mm]
horn_angle  = 20.0  # horn opening half-angle in both x and y [degrees]

### Frequency
f_start = 10e9
f_stop  = 20e9
f0      = 15e9      # centre frequency of interest [Hz]

### Derived parameters
fc_wg    = C0 / (2 * a_wg * unit)                          # TE10 cutoff [Hz]
lambda_g = C0/f0/unit / np.sqrt(1 - (fc_wg/f0)**2)        # guided wavelength at f0 [mm]

back_short = lambda_g / 4       # quarter-wave back-short [mm]
pin_length = b_wg * 0.55        # probe length [mm]  ← tune for best impedance match
pin_r      = 0.5                # probe cross-section half-width [mm] (≈ SMA inner conductor)
port_h     = 1.0                # lumped-port gap height at pin base [mm]
feed_R     = 50.0               # feed resistance [Ohm]

z_back = -feed_length           # inner face of the back-short wall [mm]
z_pin  = z_back + back_short    # pin position along the waveguide axis [mm]

# Horn aperture half-dimensions (at z = horn_length)
horn_ax = a_wg/2 + np.sin(np.deg2rad(horn_angle)) * horn_length
horn_ay = b_wg/2 + np.sin(np.deg2rad(horn_angle)) * horn_length

# Aperture area for antenna efficiency calculation
A_aperture = (2*horn_ax*unit) * (2*horn_ay*unit)

print(f'TE10 cutoff:            {fc_wg/1e9:.2f} GHz')
print(f'Guided wavelength at f0:{lambda_g:.2f} mm')
print(f'Back-short distance:    {back_short:.2f} mm')
print(f'Pin z-position:         {z_pin:.2f} mm  (from back wall: {back_short:.2f} mm)')
print(f'Pin length:             {pin_length:.2f} mm  (tune for matching)')
print(f'Pin cross-section:      {2*pin_r:.1f} x {2*pin_r:.1f} mm')

### FDTD setup
FDTD = openEMS(EndCriteria=1e-4)
FDTD.SetGaussExcite(0.5*(f_start+f_stop), 0.5*(f_stop-f_start))
FDTD.SetBoundaryCond(['PML_8']*6)

### Geometry and mesh
CSX = ContinuousStructure()
FDTD.SetCSX(CSX)
mesh = CSX.GetGrid()
mesh.SetDeltaUnit(unit)

max_res   = C0/f_stop/unit/20       # ≈ lambda/20 at highest frequency ≈ 0.75 mm

# Simulation box dimensions: structure edge + lambda/4 air gap + 8-cell PML estimate
lam0      = C0/f0/unit            # free-space wavelength at f0 [mm]
bc_offset = 9                     # PML_8 + 1 cell, matches CreateNF2FFBox placement
margin    = lam0/2 + bc_offset*max_res   # lambda/2 air gap + PML overhead [mm]

x_max     = horn_ax + margin      # x boundary (same margin used for y)
y_max     = horn_ay + margin      # y boundary
z_sim_back  = lam0/4 + bc_offset*max_res  # lambda/4 free space + PML overhead behind back wall
z_sim_front = margin              # in front of horn aperture: lambda/2 + PML

# Fixed mesh lines at key geometry locations
mesh.AddLine('x', [-x_max, -horn_ax, -(a_wg/2 + wg_t), -a_wg/2, -pin_r, pin_r,
                   a_wg/2, a_wg/2 + wg_t, horn_ax, x_max])
mesh.AddLine('y', [-y_max, -horn_ay, -(b_wg/2 + wg_t), -b_wg/2,
                   -b_wg/2+port_h, -b_wg/2+pin_length,
                   b_wg/2, b_wg/2 + wg_t, horn_ay, y_max])
mesh.AddLine('z', [z_back - z_sim_back, z_back - lam0/4, z_back - wg_t, z_back,
                   z_pin-pin_r, z_pin+pin_r,
                   0, horn_length, horn_length + z_sim_front])

### Create horn antenna geometry
horn = CSX.AddMetal('horn')

## Feed waveguide walls (closed rectangular tube from z_back to z=0)
# left wall  (at x = -a_wg/2)
horn.AddBox(priority=10, start=[-a_wg/2 - wg_t, -b_wg/2, z_back], stop=[-a_wg/2, b_wg/2, 0])
# right wall (at x = +a_wg/2)
horn.AddBox(priority=10, start=[ a_wg/2, -b_wg/2, z_back], stop=[ a_wg/2 + wg_t, b_wg/2, 0])
# top wall   (at y = +b_wg/2)
horn.AddBox(priority=10, start=[-a_wg/2 - wg_t,  b_wg/2,        z_back], stop=[ a_wg/2 + wg_t,  b_wg/2 + wg_t, 0])
# bottom wall (at y = -b_wg/2)
horn.AddBox(priority=10, start=[-a_wg/2 - wg_t, -b_wg/2 - wg_t, z_back], stop=[ a_wg/2 + wg_t, -b_wg/2,        0])

## Back-short wall (metal lid closing the waveguide)
horn.AddBox(priority=10,
            start=[-a_wg/2 - wg_t, -b_wg/2 - wg_t, z_back - wg_t],
            stop= [ a_wg/2 + wg_t,  b_wg/2 + wg_t,  z_back       ])

## Flared horn walls (four trapezoidal plates, one per side)
#
# LinPoly coordinate convention (from CSXCAD C++ cyclic index formula):
#   norm_dir='y': points[0] = z-coords, points[1] = x-coords
#   norm_dir='x': points[0] = y-coords, points[1] = z-coords
#
# Each plate starts in the plane of its normal direction at y/x = 0,
# is extruded by wg_t (centred on the elevation), then transformed with
# a rotation (to create the flare angle) and a translation (to the waveguide edge).

# Shared z-coordinates for the top/bottom horn-wall polygon
z_tb = np.array([0, horn_length, horn_length, 0])
# Corresponding x-coordinates (trapezoid widening along z)
x_tb = np.array([ a_wg/2, a_wg/2 + np.sin(np.deg2rad(horn_angle))*horn_length,
                 -a_wg/2 - np.sin(np.deg2rad(horn_angle))*horn_length, -a_wg/2])

# Bottom horn wall: rotate +horn_angle around x → far end tilts to −y
p = horn.AddLinPoly(points=[z_tb, x_tb], norm_dir='y',
                    elevation=-wg_t/2, length=wg_t, priority=10)
p.AddTransform('RotateAxis', 'x',  horn_angle)
p.AddTransform('Translate', [0, -b_wg/2 - wg_t/2, 0])

# Top horn wall: rotate −horn_angle around x → far end tilts to +y
p = horn.AddLinPoly(points=[z_tb, x_tb], norm_dir='y',
                    elevation=-wg_t/2, length=wg_t, priority=10)
p.AddTransform('RotateAxis', 'x', -horn_angle)
p.AddTransform('Translate', [0,  b_wg/2 + wg_t/2, 0])

# Shared y-coordinates for the left/right horn-wall polygon
# (spans full waveguide height including wall thickness, flares outward in y)
y_lr = np.array([ b_wg/2 + wg_t, b_wg/2 + wg_t + np.sin(np.deg2rad(horn_angle))*horn_length,
                 -b_wg/2 - wg_t - np.sin(np.deg2rad(horn_angle))*horn_length, -b_wg/2 - wg_t])
z_lr = np.array([0, horn_length, horn_length, 0])

# Left horn wall: rotate −horn_angle around y → far end tilts to −x
p = horn.AddLinPoly(points=[y_lr, z_lr], norm_dir='x',
                    elevation=-wg_t/2, length=wg_t, priority=10)
p.AddTransform('RotateAxis', 'y', -horn_angle)
p.AddTransform('Translate', [-a_wg/2 - wg_t/2, 0, 0])

# Right horn wall: rotate +horn_angle around y → far end tilts to +x
p = horn.AddLinPoly(points=[y_lr, z_lr], norm_dir='x',
                    elevation=-wg_t/2, length=wg_t, priority=10)
p.AddTransform('RotateAxis', 'y',  horn_angle)
p.AddTransform('Translate', [ a_wg/2 + wg_t/2, 0, 0])

### Coaxial pin feed
# The feed is split into two parts that together model a coaxial probe:
#
#   ┌──────┐  ← pin tip      (y = −b_wg/2 + pin_length)
#   │ PEC  │  ← metal pin body (CSX metal box, cross-section 2*pin_r × 2*pin_r)
#   │ pin  │
#   └──────┘  ← port top     (y = −b_wg/2 + port_h)
#   [lumped]  ← short port gap modelling the coaxial connector transition
#  ══════════ ← bottom wall   (y = −b_wg/2)

# Short lumped port (the coaxial connector gap, same cross-section as pin body)
start = [-pin_r, -b_wg/2,          z_pin - pin_r]
stop  = [ pin_r, -b_wg/2 + port_h, z_pin + pin_r]
port = FDTD.AddLumpedPort(1, feed_R, start, stop, 'y', 1.0, priority=5)

# Metal pin body (PEC post from port top to pin tip, realistic cross-section)
pin_metal = CSX.AddMetal('pin')
pin_metal.AddBox(priority=10,
                 start=[-pin_r, -b_wg/2 + port_h,  z_pin - pin_r],
                 stop= [ pin_r, -b_wg/2 + pin_length, z_pin + pin_r])

### Smooth mesh and create NF2FF recording box
mesh.SmoothMeshLines('all', max_res, 1.4)
nf2ff = FDTD.CreateNF2FFBox()

### Optional: write XML and launch AppCSXCAD for geometry inspection
if 1:
    CSX_file = os.path.join(Sim_Path, 'horn_pinfeed.xml')
    if not os.path.exists(Sim_Path):
        os.mkdir(Sim_Path)
    CSX.Write2XML(CSX_file)
    from CSXCAD import AppCSXCAD_BIN
    os.system(AppCSXCAD_BIN + ' "{}"'.format(CSX_file))

### Run the simulation
if not post_proc_only:
    FDTD.Run(Sim_Path, cleanup=True)

### Post-processing
freq = np.linspace(f_start, f_stop, 401)
port.CalcPort(Sim_Path, freq)

s11    = port.uf_ref / port.uf_inc
s11_dB = 20.0 * np.log10(np.abs(s11))
Zin    = port.uf_tot / port.if_tot

## Reflection coefficient S11
fig, ax = plt.subplots(num='S11', tight_layout=True)
ax.plot(freq/1e9, s11_dB, 'k-', linewidth=2)
ax.axhline(-10, color='gray', linestyle='--', linewidth=1)
ax.grid()
ax.set_xmargin(0)
ax.set_ylim([-40, 5])
ax.set_xlabel('Frequency (GHz)')
ax.set_ylabel('S11 (dB)')
ax.set_title('Reflection Coefficient S11')

## Input impedance
fig, ax = plt.subplots(num='Zin', tight_layout=True)
ax.plot(freq/1e9, np.real(Zin), 'k-',  linewidth=2, label=r'$\Re\{Z_{in}\}$')
ax.plot(freq/1e9, np.imag(Zin), 'r--', linewidth=2, label=r'$\Im\{Z_{in}\}$')
ax.grid()
ax.set_xmargin(0)
ax.set_xlabel('Frequency (GHz)')
ax.set_ylabel('Impedance (Ohm)')
ax.set_title('Input Impedance')
ax.legend()

## Far-field radiation pattern at f0
theta = np.arange(-180.0, 180.0, 2.0)
phi   = [0.0, 90.0]
print(f'Calculating far field at f0 = {f0/1e9:.1f} GHz ...')
nf2ff_res = nf2ff.CalcNF2FF(Sim_Path, f0, theta, phi)

Dmax_dBi = 10.0 * np.log10(nf2ff_res.Dmax[0])
G_a = 4 * np.pi * A_aperture / (C0/f0)**2      # ideal aperture gain (uniform illumination)
e_a = nf2ff_res.Dmax[0] / G_a                  # aperture efficiency

print(f'Directivity:          Dmax = {Dmax_dBi:.1f} dBi')
print(f'Aperture efficiency:  e_a  = {e_a*100:.1f} %')

E_norm = (20.0*np.log10(nf2ff_res.E_norm[0] / np.max(nf2ff_res.E_norm[0]))
          + Dmax_dBi)

fig, ax = plt.subplots(num='Pattern', tight_layout=True)
ax.plot(theta, E_norm[:, 0], 'k-',  linewidth=2, label='E-plane  (phi=0°)')
ax.plot(theta, E_norm[:, 1], 'r--', linewidth=2, label='H-plane  (phi=90°)')
ax.grid()
ax.set_xmargin(0)
ax.set_xlabel('Theta (deg)')
ax.set_ylabel('Directivity (dBi)')
ax.set_title(f'Far-Field Pattern at {f0/1e9:.1f} GHz')
ax.legend()

plt.show()
